carl/envs/rna/rna_environment.py

```diff
@@ -0,0 +1,352 @@
+        Returns:
+            The site that pairs with <site> if exists.
+        """
+        return self._pairing_encoding[site]
+
+
+class _Design(object):
+    """
+    Class of the designed candidate solution.
+    """
+
+    action_to_base = {0: "G", 1: "A", 2: "U", 3: "C"}
+    action_to_pair = {0: "GC", 1: "CG", 2: "AU", 3: "UA"}
+
+    def __init__(self, length=None, primary=None):
+        """
+        Initialize a candidate solution.
+        Args:
+            length: The length of the candidate solution.
+            primary: The sequence of the candidate solution.
+        """
+        if primary:
+            self._primary_list = primary
+        else:
+            self._primary_list = [None] * length
+        self._dot_bracket = None
+        self._current_site = 0
+
+    def get_mutated(self, mutations, sites):
+        """
+        Locally change the candidate solution.
+        Args:
+            mutations: Possible mutations for the specified sites
+            sites: The sites to be mutated
+        Returns:
+            A Design object with the mutated candidate solution.
+        """
+        mutatedprimary = self._primary_list.copy()
+        for site, mutation in zip(sites, mutations):
+            mutatedprimary[site] = mutation
+        return _Design(primary=mutatedprimary)
+
+    def assign_sites(self, action, site, paired_site=None):
+        """
+        Assign nucleotides to sites for designing a candidate solution.
+        Args:
+            action: The agents action to assign a nucleotide.
+            site: The site to which the nucleotide is assigned to.
+            paired_site: defines if the site is assigned with a base pair or not.
+        """
+        self._current_site += 1
+        if paired_site:
+            base_current, base_paired = self.action_to_pair[action]
+            self._primary_list[site] = base_current
+            self._primary_list[paired_site] = base_paired
+        else:
+            self._primary_list[site] = self.action_to_base[action]
+
+    @property
+    def first_unassigned_site(self):
+        try:
+            while self._primary_list[self._current_site] is not None:
+                self._current_site += 1
+            return self._current_site
+        except IndexError:
+            return None
+
+    @property
+    def primary(self):
+        return "".join(self._primary_list)
+
+
+def _random_epoch_gen(data):
+    """
+    Generator to get epoch data.
+    Args:
+        data: The targets of the epoch
+    """
+    while True:
+        for i in np.random.permutation(len(data)):
+            yield data[i]
+
+
+@dataclass
+class EpisodeInfo:
+    """
+    Information class.
+    """
+
+    __slots__ = ["target_id", "time", "normalized_hamming_distance"]
+    target_id: int
+    time: float
+    normalized_hamming_distance: float
+
+
+class RnaDesignEnvironment(gym.Env):
+    """
+    The environment for RNA design using deep reinforcement learning.
```

Comment:
Fix docstring
(I know that they were not nice before)

Reply:
Especially for this class.
Please describe goal/what this env does in general, states, actions, rewards, starting state, termination criteria

Reply:
Done